### PR TITLE
[Merged by Bors] - feat(Rat): Numerator and denominator of `q ^ n`

### PR DESCRIPTION
**Pull request by YaelDillies**
Prove `(q ^ n).num = q.num ^ n` and `(q ^ n).den = q.den ^ n` by making those defeq. It is somewhat painful. `(q ^ n).den = q.den ^ n` is also defeq for `NNRat`, but not `(q ^ n).num = q.num ^ n` due to the `Int.natAbs` in the definition of `NNRat.num`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Rat/Defs.lean [hunk 1]
```diff
@@ -216,12 +218,38 @@ lemma divInt_mul_divInt' (n₁ d₁ n₂ d₂ : ℤ) : (n₁ /. d₁) * (n₂ /.
...
+  pow q n := ⟨q.num ^ n, q.den ^ n, by simp [Nat.pow_eq_zero], by
+    rw [Int.natAbs_pow]; exact q.reduced.pow _ _⟩
+
+lemma pow_def (q : ℚ) (n : ℕ) : q ^ n = ⟨q.num ^ n, q.den ^ n, by simp [Nat.pow_eq_zero], by
+    rw [Int.natAbs_pow]; exact q.reduced.pow _ _⟩ := rfl
+
+lemma pow_eq_mkRat (q : ℚ) (n : ℕ) : q ^ n = mkRat (q.num ^ n) (q.den ^ n) := by
+  rw [pow_def, mk_eq_mkRat]
+
+lemma pow_eq_divInt (q : ℚ) (n : ℕ) : q ^ n = q.num ^ n /. q.den ^ n := by
```
Comment (eric-wieser) [inline Mathlib/Data/Rat/Defs.lean:243]: This doesn't seem to match the `mul` lemmas; should we have `divInt_pow` instead to match `divInt_mul_divInt'`?

**Review by eric-wieser — APPROVED** [review]
Nice, this probably makes `^` on `Rat` computationally faster too.

File: Mathlib/Data/Rat/Defs.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Data/Rat/Defs.lean:243]: This matches `Rat.mul_def'` (which is getting renamed to a matching name in #12505). But also I was trying to prove `divInt_pow` and failed. I can try again.

File: Mathlib/Data/Rat/Defs.lean [hunk 2]
```diff
@@ -227,6 +235,26 @@ alias divInt_eq_divInt := divInt_eq_iff
 @[deprecated] alias mul_num_den := mul_eq_mkRat
 #align rat.mul_num_denom Rat.mul_eq_mkRat
 
+instance instPowNat : Pow ℚ ℕ where
+  pow q n := ⟨q.num ^ n, q.den ^ n, by simp [Nat.pow_eq_zero], by
+    rw [Int.natAbs_pow]; exact q.reduced.pow _ _⟩
+
+lemma pow_def (q : ℚ) (n : ℕ) : q ^ n = ⟨q.num ^ n, q.den ^ n, by simp [Nat.pow_eq_zero], by
+    rw [Int.natAbs_pow]; exact q.reduced.pow _ _⟩ := rfl
```
Comment (eric-wieser) [inline Mathlib/Data/Rat/Defs.lean:243]: super nit: I think this indent makes more sense
```suggestion
lemma pow_def (q : ℚ) (n : ℕ) :
  q ^ n = ⟨q.num ^ n, q.den ^ n, by simp [Nat.pow_eq_zero], by
    rw [Int.natAbs_pow]; exact q.reduced.pow _ _⟩ := rfl
```

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks!

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge

**Comment by eric-wieser** [comment]
I think the build is still failing?

bors r-

**Comment by YaelDillies** [comment]
Sorry, the mobile app is very limiting in what I can see 🙈

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**